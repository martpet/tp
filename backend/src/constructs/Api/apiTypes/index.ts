export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
