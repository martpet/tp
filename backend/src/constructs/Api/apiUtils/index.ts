export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
