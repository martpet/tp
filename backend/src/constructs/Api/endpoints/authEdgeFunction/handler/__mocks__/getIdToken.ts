export const getIdToken = vi.fn().mockResolvedValue('dummyIdToken');
