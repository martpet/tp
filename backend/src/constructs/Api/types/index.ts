export * from './EndpointCookies';
export * from './EndpointEnvVars';
export * from './EndpointHeaders';
export * from './EndpointsQueryStrings';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
