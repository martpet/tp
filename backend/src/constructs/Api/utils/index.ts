export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
