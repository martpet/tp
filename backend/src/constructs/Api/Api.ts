import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/authEdgeFunction/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'HttpApi');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'Alias', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
