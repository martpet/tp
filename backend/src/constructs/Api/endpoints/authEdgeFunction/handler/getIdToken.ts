import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils/getIdTokenPayload';
import { refreshTokenExpiredErrorMessage, region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp, aud: clientId } = await getIdTokenPayload(idToken);

  const idTokenExpires = (exp - 5) * 1000; // 'exp' is in seconds since epoch

  if (idTokenExpires > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error(refreshTokenExpiredErrorMessage);
};
