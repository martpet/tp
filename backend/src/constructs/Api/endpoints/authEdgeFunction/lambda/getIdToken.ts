import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils';
import { region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp, aud: clientId } = getIdTokenPayload(idToken);
  const idTokenExpires = exp - 5000;

  if (idTokenExpires > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error('"id" and "refresh" tokens have expired');
};
