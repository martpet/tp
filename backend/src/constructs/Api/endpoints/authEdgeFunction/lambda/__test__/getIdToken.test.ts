import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import millis from 'milliseconds';

import { IdTokenPayload } from '~/constructs/Api/types';
import { getIdTokenPayload } from '~/constructs/Api/utils';

import { fetchNewIdToken } from '../fetchNewIdToken';
import { getIdToken } from '../getIdToken';

vi.mock('../fetchNewIdToken');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');

const ddbMock = mockClient(DynamoDBDocumentClient);
const originalDate = Date.now();
const idTokenExpires = Date.now() + millis.days(1);
const refreshTokenExpires = Date.now() + millis.years(1);
const args = ['dummySessionId'] as Parameters<typeof getIdToken>;

vi.mocked(fetchNewIdToken).mockResolvedValue('newDummyIdToken');

vi.mocked(getIdTokenPayload).mockReturnValue({
  exp: idTokenExpires,
  aud: 'dummyAud',
} as IdTokenPayload);

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(GetCommand).resolves({
    Item: {
      idToken: 'dummyIdToken',
      refreshToken: 'dummyRefreshToken',
      refreshTokenExpires,
    },
  });
});

describe('getTokens', () => {
  it('sends "GetCommand" to DynamoDB with correct args', async () => {
    await getIdToken(...args);
    expect(ddbMock.commandCalls(GetCommand)[0].args[0].input).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(getIdToken(...args)).resolves.toMatchSnapshot();
  });

  describe('when "Item" prop is missing from "GetCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(GetCommand).resolves({});
    });

    it('rejects with a correct value', () => {
      return expect(getIdToken(...args)).rejects.toMatchSnapshot();
    });
  });

  describe('when "idToken" has expired', () => {
    beforeAll(() => {
      vi.setSystemTime(new Date(idTokenExpires + millis.days(1)));
    });

    afterAll(() => {
      vi.setSystemTime(originalDate);
    });

    it('calls "fetchNewIdToken" with correct args', async () => {
      await getIdToken(...args);
      expect(vi.mocked(fetchNewIdToken).mock.calls).toMatchSnapshot();
    });

    it('resolves with a correct value', () => {
      return expect(getIdToken(...args)).resolves.toMatchSnapshot();
    });
  });

  describe('when "refreshToken" has expired', () => {
    beforeAll(() => {
      vi.setSystemTime(new Date(refreshTokenExpires + millis.days(1)));
    });

    afterAll(() => {
      vi.setSystemTime(originalDate);
    });

    it('rejects with a correct value', () => {
      return expect(getIdToken(...args)).rejects.toMatchSnapshot();
    });
  });
});
