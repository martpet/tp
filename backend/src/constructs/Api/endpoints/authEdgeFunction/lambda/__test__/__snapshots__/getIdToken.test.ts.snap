// Vitest Snapshot v1

exports[`getTokens > resolves with a correct value 1`] = `"dummyIdToken"`;

exports[`getTokens > sends "GetCommand" to DynamoDB with correct args 1`] = `
{
  "Key": {
    "id": "dummySessionId",
  },
  "TableName": "TripPics-Sessions",
}
`;

exports[`getTokens > when "Item" prop is missing from "GetCommand" output > rejects with a correct value 1`] = `[Error: could not find session with id "dummySessionId"]`;

exports[`getTokens > when "idToken" has expired > calls "fetchNewIdToken" with correct args 1`] = `
[
  [
    {
      "clientId": "dummyAud",
      "refreshToken": "dummyRefreshToken",
      "sessionId": "dummySessionId",
    },
  ],
]
`;

exports[`getTokens > when "idToken" has expired > resolves with a correct value 1`] = `"newDummyIdToken"`;

exports[`getTokens > when "refreshToken" has expired > rejects with a correct value 1`] = `[Error: Refresh token expired]`;
